### ISSUE TITLE
bug report: Dovecot compression not working

### ISSUE THREAD

**Issue by alex9434** [docker-mailserver/docker-mailserver#4178]
### 📝 Preliminary Checks

- [X] I tried searching for an existing issue and followed the [debugging docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/) advice, but still need assistance.

### 👀 What Happened?

I have the following setting in `dovecot.cf`

```
protocol imap {
	mail_plugins = $mail_plugins zlib imap_zlib
}

# Enable these only if you want compression while saving:
plugin {
	zlib_save = gz
	zlib_save_level = 9
}

plugin {
	## NOTE! Setting was called imap_zlib_compression_level before 2.3.15
	## imap_compress_<algorithm>_level = value
	##
	imap_compress_deflate_level = 6
}
```

### 👟 Reproduction Steps

New mail that arrives in maildir is does not get compressed. These settings should save all new email compressed with gz.

### 🐋 DMS Version

v14.0.0

### 💻 Operating System and Architecture

Debian 11 x86_64

### ⚙️ Container configuration files

```yml
services:
  mailserver:
    image: ghcr.io/docker-mailserver/docker-mailserver:latest
    read_only: false
    container_name: mailserver
    # Provide the FQDN of your mail server here (Your DNS MX record should point to this value)
    hostname: mail.emimx.com
    environment:
      - SSL_TYPE=manual
      - SSL_CERT_PATH=/tmp/dms/custom-certs/fullchain.cer
      - SSL_KEY_PATH=/tmp/dms/custom-certs/cert.key   
    env_file: mailserver.env
    # More information about the mail-server ports:
    # https://docker-mailserver.github.io/docker-mailserver/latest/config/security/understanding-the-ports/
    ports:
      - "25:25"    # SMTP  (explicit TLS => STARTTLS, Authentication is DISABLED => use port 465/587 instead)
      - "143:143"  # IMAP4 (explicit TLS => STARTTLS)
      - "465:465"  # ESMTP (implicit TLS)
      - "587:587"  # ESMTP (explicit TLS => STARTTLS)
      - "993:993"  # IMAP4 (implicit TLS)
    volumes:
      - ./mail/:/var/mail/
      - ./mail-state/:/var/mail-state/
      - ./mail-logs/:/var/log/mail/
      - ./config/:/tmp/docker-mailserver/
      - ./letsencrypt:/etc/letsencrypt
      - /etc/localtime:/etc/localtime:ro
      - /etc/ssl/private/cert.ecc/:/tmp/dms/custom-certs/:ro
    restart: unless-stopped
    stop_grace_period: 1m
    # Uncomment if using `ENABLE_FAIL2BAN=1`:
    # cap_add:
    #   - NET_ADMIN
    healthcheck:
      test: "ss --listening --tcp | grep -P 'LISTEN.+:smtp' || exit 1"
      timeout: 3s
      retries: 0
```

### 📜 Relevant log output

```Text
new email stored uncompressed
```

**Comment by polarathene** [comment]
We do not provide support for features not supported by DMS. Custom config with `dovecot.cf` is supported and known to work.

Ensure you actually followed the [troubleshooting guide docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/debugging/), especially about how to restart DMS properly when updating config. Beyond that follow the advice on logging to provide more information and try look into using the `doveconf` command within the container to see if your config was applied (_if it was, you'll need to investigate why it's not having an effect_).

**Comment by alex9434** [comment]
Of course I have deleted the container before restarting. The configuration is included in doveconfig.
```
$ docker exec -ti mailserver doveconf | grep zlib
  zlib_save = gz
  zlib_save_level = 9
  mail_plugins = " quota imap_quota zlib imap_zlib"
```
Since zlib is included in the mailserver image, I think this feature is supported by DMS. Unfortunately it does not seem to work. How can I debug why the emails are not compressed on saving. Are they delivered into the maildir by dovecot-lda or by postfix?

**Comment by polarathene** [comment]
> Unfortunately it does not seem to work. How can I debug why the emails are not compressed on saving.

If it's not documented in our docs, it's not something we can offer support for.

You can try creating the project discussions page to ask the community there, but you'll most likely be troubleshooting on your own for some time.

---

> Are they delivered into the maildir by dovecot-lda or by postfix?

With DMS Postfix is configured to deliver mail to Dovecot over LMTP. LDA is not used.

**Comment by alex9434** [comment]
> With DMS Postfix is configured to deliver mail to Dovecot over LMTP. LDA is not used.
That explains it. Compression can only work if dovecot-lda takes mail from Postfix and delivers it to a user’s mailbox (https://doc.dovecot.org/2.3/configuration_manual/protocols/lda/)

May I suggest to use dovecot-lda for delivery? That allows not only compression but improves mailbox indexing for faster notification, allows additional mailbox formats, etc. What was the reason not to use dovecot-lda? It seems to be recommended by dovecot?

**Comment by georglauterbach** [comment]
> > With DMS Postfix is configured to deliver mail to Dovecot over LMTP. LDA is not used.
> 
> That explains it. Compression can only work if dovecot-lda takes mail from Postfix and delivers it to a user’s mailbox (https://doc.dovecot.org/2.3/configuration_manual/protocols/lda/)
> 
> May I suggest to use dovecot-lda for delivery? That allows not only compression but improves mailbox indexing for faster notification, allows additional mailbox formats, sieve support, better quota enforcement etc. What was the reason not to use dovecot-lda? It seems to be recommended by dovecot.

I guess the original author of this image has opted to go with LMTP. If you can provide a set of changes and rationale to switch to LDA, we can try. But maintainers have little time; changing a feature that works well currently is not something we as maintainers can afford,

**Comment by polarathene** [comment]
**TL;DR:**
- This is extremely likely to be misconfiguration by @alex9434 , it's fairly common when users try to customize DMS like this. Often they've not actually followed our troubleshooting docs properly, nor provided sufficient / accurate information to verify. For example the statement that LDA is required appears invalid?
- I am extremely doubtful anyone is going to contribute easy config support for LDA in DMS when it'll be easier for the user to use the existing support we already offer to write a few extra lines instead of writing more complex functionality in our startup scripts + docs + tests.
- I could take time to investigate and more than likely prove this works fine in DMS, but I have other tasks to prioritize.

---

> I guess the original author of this image has opted to go with LMTP.

I do recall looking into the switch in the past, and I'm sure I've documented it around here somewhere a few times. Found a brief mention [here](https://github.com/docker-mailserver/docker-mailserver/issues/3557):

> April 2016 - DMS https://github.com/docker-mailserver/docker-mailserver/pull/123, which introduced various configs referencing the docker user and group:
> 
> - [Dovecot UserDB auth socket](https://github.com/docker-mailserver/docker-mailserver/blob/89cb6d85b90f6070d53c15e455ebee916c899c7a/target/dovecot/auth-passwdfile.inc#L18)
>   - This doesn't seem relevant anymore in DMS, it was originally for supporting [Dovecot LDA](https://doc.dovecot.org/configuration_manual/protocols/lda/?highlight=lda#virtual-users).. but https://github.com/docker-mailserver/docker-mailserver/pull/360 to [Dovecot LMTP](https://doc.dovecot.org/configuration_manual/protocols/lmtp_server/#lmtp-server) in Oct 2016.

That then links to #360 where I can see a few linkbacks from myself and [the discussion / reasoning for switching from LDA to LMTP](https://github.com/docker-mailserver/docker-mailserver/issues/305).

[This beefy sized comment](https://github.com/docker-mailserver/docker-mailserver/issues/3739#issuecomment-1902490150) of mine also notes (_context from quoted content is minimized here, this is just a focused highlight but pursuing a switch to LDA would need to account for the broader context I reported_):

> - Other aliases (`postfix-virtual.cf`) test coverage:
>   - [This CC alias mail sent](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L91-L94) collides with a [similar alias test mail](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L82) where test-case would pass for either. There is no separate test coverage on the CC address. See reference section below for history details on why CC was used and what it's actually intended to test (_relevance is lost since switching from Dovecot LDA to LMTP_). [Related review comment](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438973153).

> - [Oct 2016 removed `dovecot_destination_recipient_limit=1` from Postfix `main.cf`](https://github.com/docker-mailserver/docker-mailserver/pull/363) due to ["unused parameter" warning spam in logs](https://github.com/docker-mailserver/docker-mailserver/issues/362).
>   - Earlier in Oct 2016, https://github.com/docker-mailserver/docker-mailserver/pull/360 (which was configured as the dovecot transport in `master.cf`). Hence the warning log spam issue that followed.
>   - As docs for the [Postfix `master.cf` `pipe` daemon](https://www.daemon-systems.org/man/pipe.8.html) describe under _"Single-Recipient Delivery"_, multi-recipient delivery can be problematic when `Delivered-to:` / `X-Original-To:` message headers are prepended.
>     - Previously with the `dovecot` transport in `master.cf` this used [`flags=DRhu`](https://github.com/docker-mailserver/docker-mailserver/commit/a62062c382ef47abe49acf5805b8e8ccddcf0da0#diff-382e37a673d41779bffd7dfe054feb345c1f3e5f7121698f2a52e7e5322c7169L71) which as the `pipe` daemon docs (on flags section) states, requires the limit parameter is set to `1` to correctly work.

> [Feb 2017 issue of the change to LMTP affecting Dovecot Sieve](https://github.com/docker-mailserver/docker-mailserver/issues/504), since the `X-Original-To:` header was not available, the mail that arrived for Sieve to process had already resolved the alias and lost this context that the additional header would have provided.
> - There are a few workarounds such as:

Finally, [this PR of mine](https://github.com/docker-mailserver/docker-mailserver/pull/3004) provides additional insights into motivations for LMTP from past contributors, it's usage where LDA is not viable, and some context within our own test suite.

---

Switching to LDA does not seem like the right course of action. There is a fair amount to account for and to accommodate such a change. It would be breaking and we'd more than likely get another set of users pushing for LMTP again.

You'd need to instead add some complexity to configure DMS to use LDA instead, and for that we defer to community contributions at our docs with a feature request that receives sufficient interest from a number of users before progressing from docs guide to official DMS support. We have this policy in place due to the burden of maintaining additional permutations, it adds friction to the project, often for the convenience of a small subset of DMS users.

I am aware of concerns in the past where some users wanted to leverage some headers with Dovecot Sieve that were lost from the LDA to LMTP switch, but I investigated this (_there is a Github Discussions topic and eventually I'll contribute a `compose.yaml` example or docs addition for guidance on it_).

Please note the [referenced Dovecot LDA docs link](https://doc.dovecot.org/2.3/configuration_manual/protocols/lda/) you provided begins with:

> _Nowadays you should probably use the [LMTP server](https://doc.dovecot.org/2.3/configuration_manual/protocols/lmtp_server/#lmtp-server) instead, because it’s somewhat easier to configure (especially related to permissions) and gives better performance._

Nothing on that linked LDA docs page mentions `compression`, nor did I see the [IMAP compression docs](https://doc.dovecot.org/2.3/configuration_manual/imap_compress/) or the [IMAP zlib plugin docs](https://doc.dovecot.org/2.3/settings/plugin/imap-zlib-plugin/#plugin-imap-zlib) state anything about LDA specific requirements.

> May I suggest to use dovecot-lda for delivery? That allows not only compression but improves mailbox indexing for faster notification, allows additional mailbox formats, sieve support, better quota enforcement etc. What was the reason not to use dovecot-lda? It seems to be recommended by dovecot.

Sorry to be rude, but I get a bit irritated when such statements are made when your link for justification contradicts that very statement. In fairness, it's my fault for responding with everything I did prior to verifying your statement by checking the actual link.

We already have mailbox format support, sieve, quota, etc. That should be clearly obvious from our docs. I don't think you've actually configured for LDA yourself to get the compression support?

I could side-track my current backlog (_but I won't without good reason_) to verify that compression is supported perfectly fine with LMTP, and I expect that it is. Similar to the Github Discussion I mentioned earlier, I provided plenty of insights in February for someone to implement it. Over time more users chimed in with attempts, but they had issues making it work, so 6 months later [I took the time to investigate myself](https://github.com/orgs/docker-mailserver/discussions/3898#discussioncomment-10263045) and made it work. I realize I have more expertise in the area, but that is why I'm also fairly confident with my earlier assumption of misconfiguration on your end.

**Comment by polarathene** [comment]
As this is not officially supported by DMS, I do not deem this a bug. Raise a feature request, get enough user support via 👍 reacts and ideally show a working configuration (_since if you configure for LDA it shouldn't make any actual difference vs LMTP AFAIK_).

---

Actually.... looking at the IMAP Dovecot docs I linked in the previous comment. I think @alex9434 has misunderstood what the plugin does?

> https://doc.dovecot.org/2.3/configuration_manual/imap_compress/
> 
> Dovecot supports the IMAP COMPRESS ([RFC 4978](https://datatracker.ietf.org/doc/html/rfc4978.html)) extension.

> 
> It allows an IMAP client to dynamically enable stream compression for an IMAP session.

The IMAP mail client needs to use the extension when connecting to Dovecot. IMAP is used for mail retrieval, the compression is on-demand for transfer between Dovecot and the MUA. This isn't about delivery from Postfix with storage compression on-disk.

If you want compressed storage, an agnostic approach would be to use a filesystem that supports that. BTRFS and NTFS are common examples.

---

I will close this given the above. If you feel that I am mistaken, feel free to comment on why it should be re-opened. If this feature was not misunderstood, open a feature request (_you may be waiting a long time_). If you instead just want the functionality of this Dovecot plugin (_as I've described it_), open a new discussion on our Github Discussions page to seek community support for customizing DMS when our docs are not sufficient.

**Comment by alex9434** [comment]
@polarathene 
> * This is extremely likely to be misconfiguration by @alex9434 , it's fairly common when users try to customize DMS like this. Often they've not actually followed our troubleshooting docs properly, nor provided sufficient / accurate information to verify. For example the statement that LDA is required appears invalid?

What evidence do you have that this is a misconfiguration on my part? I have shared the content of my dovecot.cf as well as the evidence that this configuration is actually picked up. What in that dovecot.cf is misconfigured? What makes you think, that I did not  follow your troubleshooting guides properly?

> Sorry to be rude, but I get a bit irritated when such statements are made when your link for justification contradicts that very statement. In fairness, it's my fault for responding with everything I did prior to verifying your statement by checking the actual link.
> 
> We already have mailbox format support, sieve, quota, etc. That should be clearly obvious from our docs. I don't think you've actually configured for LDA yourself to get the compression support?

Sorry to agree here, but I - in fact - found your statements a bit "rude". I actually have configured compression support in my previous dovecot implementation. What brings you to the conclusion that I have never done that? I have experienced space savings of around 40% with compression which is in line with observations from others that I have seen in other forums. What in the document contradicts what I have said? I quoted one part of the document and you quoted a different part. Where specifically is the contradiction?

> ## Dovecot Compression feature is not on-disk storage compression: On-the fly compression of IMAP/POP 
> Actually.... looking at the IMAP Dovecot docs I linked in the previous comment. I think @alex9434 has misunderstood what the plugin does?
Is this more a mis-understanding on **YOUR** part? Dovecot supports **TWO** types of compression: 
1) IMAP COMPRESS ([RFC 4978](https://datatracker.ietf.org/doc/html/rfc4978.html)) provided by the imap_zlib plugin. imap_zlib accelerates delivery over slow connection (e.g. mobile) when the client understands the extension.
2) Compressed storage (for that the dovecot LDA needs to be invoked, Postfix does not have that capability) provided by the zlib plugin. zlib saves space and speeds up the compressed delivery by avoiding on-the-fly compression because the mails are already stored compressed.

This is why I also  have tried to involve the TWO relevant plugins  statements in the provided dovecot.cf
```
protocol imap {
	mail_plugins = $mail_plugins zlib imap_zlib
}
```
imap_zlib: https://doc.dovecot.org/2.3/configuration_manual/imap_compress/#imap-compress
zlib: https://doc.dovecot.org/2.3/configuration_manual/zlib_plugin/

Closing - Proposal for proceeding
Does it really make sense to open a "feature request"? Or is this a futile exercise? Given the tonality/content of your response it probably makes more sense to a) short term live with the uncompressed maildir, b) middle term write a separate batch file for overnight compression (unfortunately additional write cycles for the SSD), and c) long term search for a different email solution.

Independent of my requirement, storage compression seems to be an obvious benefit given the a) high compressibility of email and b) the potential speed improvements for compressed delivery.

**Comment by georglauterbach** [comment]
No need for this to become heated :) We will have a look at [your latest comments](https://github.com/docker-mailserver/docker-mailserver/issues/4178#issuecomment-2376712464) again re-evaluate. If it turns out that we misunderstood something, we will reconsider. Please do not open a feature request yet; I will re-open this first and we will decide here. On-Disk compression might be worthwhile, but I'd like for @polarathene to have a look at your feedback here because I don't know the details.

**Comment by polarathene** [comment]
> What evidence do you have that this is a misconfiguration on my part?
> - I have shared the content of my `dovecot.cf` as well as the evidence that this configuration is actually picked up. What in that `dovecot.cf is` misconfigured?
> - What makes you think, that I did not follow your troubleshooting guides properly?

Given that the documented functionality of the plugin I linked to seemed to different from what you thought it was, and the link you provided as evidence for LDA being required with LMTP being incompatible when there's nothing to say that is the case... It would suggest that you're not reading those docs properly. (_**EDIT:** As conveyed below, I was not aware at the time of a similar plugin for on-disk compression_)

If you're not reading those properly, and as maintainers it's quite common for our users to misconfigure or claim they've done something but it turns out they didn't do it properly, I would be inclined to assumed that it's likely not an issue with DMS but the user. Especially with the LDA will fix it claim when there is no verification of proof on your end to say otherwise.

---

> Sorry to agree here, but I - in fact - found your statements a bit "rude".

I realize they were rude, that was the whole point of acknowledging how it would come across 😕 Communication skills are not something I excel at.

> - What in the document contradicts what I have said?

You linked to [this docs page](https://doc.dovecot.org/2.3/configuration_manual/protocols/lda/) with the statement that LMTP was not compatible with the feature you wanted, that it would only work with LDA. You then made a statement about LDA benefits over LMTP that were not true.

>  I quoted one part of the document and you quoted a different part. Where specifically is the contradiction?

Sorry, I do not see what you're referring to as a quote from past comments? You did not have any `>` markdown quote syntax used, could you be more specific please?

---

> I actually have configured compression support in my previous dovecot implementation.
> - What brings you to the conclusion that I have never done that?
> - I have experienced space savings of around 40% with compression which is in line with observations from others that I have seen in other forums.

I was not stating you had never done this or that on-disk compression was not possible. Just the information I had come across was not matching what you seemed to be wanting.

I ended my previous comment acknowledging that I may have missed something and been mistaken.

> Is this more a mis-understanding on **YOUR** part? Dovecot supports **TWO** types of compression

Yes, you are correct, thank you.

I was mistaken as I was irritated putting in the time to assess your request for LDA being needed, only to find no evidence that such was required. On that page I searched for compression, found the sidebar link and the zlib imap plugin it refers to. I should have also done a search on zlib to find the other functionality, or pay more attention to the initial config snippet you provided compared to the one from the docs I referenced.

I hope that you can understand that I am very busy, so the less work I have to do to find the information that is relevant, the more you can help us both save time, and when you instead provide me with information that wastes my time to assist you I will understandably become frustrated / rude (_this is my time I volunteer for your benefit, not the other way around_).

---

> 2\. Compressed storage (for that the dovecot LDA needs to be invoked, Postfix does not have that capability) provided by the zlib plugin.

From the other zlib docs link you provided:

> https://doc.dovecot.org/2.3/configuration_manual/zlib_plugin/
> 
> Zlib plugin can be used to read compressed mbox, maildir or dbox files. It can also be used to write (via IMAP, [LDA](https://doc.dovecot.org/2.3/configuration_manual/protocols/lda/#lda) and/or [**LMTP** Server](https://doc.dovecot.org/2.3/configuration_manual/protocols/lmtp_server/#lmtp-server)) compressed messages to [dbox](https://doc.dovecot.org/2.3/admin_manual/mailbox_formats/dbox/#dbox-mbox-format) or [Maildir](https://doc.dovecot.org/2.3/admin_manual/mailbox_formats/maildir/#maildir-mbox-format) mailboxes.

It explicitly states that LMTP is supported. Where are you getting this information that LDA must be used instead of LMTP?

You have not actually changed the configuration to LDA yourself to verify this claim have you? You are requesting a change to LDA that I am quite reluctant to accept or support, as was covered in my prior response.

---

> Does it really make sense to open a "feature request"? Or is this a futile exercise?

If you expect others to do the work for you, then yes that is one of the options available to you. It may not be a "feature" with specialized DMS config support, it may just be a documented example / guide that is added to our own docs for users.

Regardless, you are asking for something that DMS does not officially document support for. This project integrates and manages a variety of mail services, and we have tests in place for what we do support. It would be unreasonable for us to cover every permutation, hence the features like `dovecot.cf` for you to leverage when you need customizations that we don't offer out of the box.

Sometimes those require additional changes, but that is not our responsibility to figure out for you. We are not free tech support, we maintain this project for the benefit of everyone at the expense of our own time. I have received $0 over the significant time I've invested into the project since early 2021, it is a voluntary basis where I like to help but I decide where my time is spent on the project.

> Given the tonality/content of your response

I realize you do not like it, we've both made mistakes in the above discourse. I have a lot to juggle (not just DMS), and a lot to write in responses like this, taking time to make it more professional is not something I can afford for your benefit right now.

I will prioritize being critical for the best interests of this project given it's history of contributions in the past that were misinformed and led to problems I've often resolved. Unfortunately that stance does not always go down well, especially in written communications.

We do value user contributions and reports, and I do apologize for my own mistake on not being thorough enough earlier, but as this is not something with any official support in DMS, the responsibility of troubleshooting to identify where a potential bug may be, or what the correct configuration is largely falls on to you for choosing to add custom configuration.

That is not to say you will not receive any further assistance, but without sufficient user demand or further insights, this is a niche concern for us to resolve, and thus low on our priorities.

> long term search for a different email solution.

You are more than welcome to seek out alternatives that may better accommodate the features you need. MailCow and Stalwart (recommended) are quite popular choices to migrate to, you may also want to consider Mailu.

---

> Independent of my requirement, storage compression seems to be an obvious benefit given the a) high compressibility of email and b) the potential speed improvements for compressed delivery.

I've not disputed this. Although it's not something that would always be desirable to have that Dovecot feature enabled, so if you'd like to see official support via an opt-in ENV toggle please raise a feature request.

**Comment by polarathene** [comment]
As assumed user attempted to apply custom configuration and did so incorrectly. Closing again as not a bug.

## User misconfiguration

> What evidence do you have that this is a misconfiguration on my part?
> What in that `dovecot.cf` is misconfigured?

> ```
> protocol imap {
> 	mail_plugins = $mail_plugins zlib imap_zlib
> }
> ```

This was your problem. `imap_zlib` is only relevant to the IMAP protocol, while the `zlib` plugin should be available to the Dovecot protocol(s) for receiving mail (such as LMTP).

The [`zlib` plugin docs](https://doc.dovecot.org/2.3/configuration_manual/zlib_plugin/#zlib-plugin) do not scope to a specific protocol in their example. While the [`imap_zlib` plugin docs](https://doc.dovecot.org/2.3/configuration_manual/imap_compress/) do specifically use the imap protocol section for the example there. Easy mistake.

---

> The configuration is included in dovecot config.
> 
> ```console
> $ docker exec -it mailserver doveconf | grep zlib
>   zlib_save = gz
>   zlib_save_level = 9
>   mail_plugins = " quota imap_quota zlib imap_zlib"
> ```

That did not properly check the config output for the `mail_plugins` setting:

```console
# Filter config output by setting and include the line before and after the match(es):
$ doveconf | grep -B1 -A1 mail_plugins
mail_plugin_dir = /usr/lib/dovecot/modules
mail_plugins = " quota"
mail_prefetch_count = 0
--
protocol lmtp {
  mail_plugins = " quota sieve"
}
protocol lda {
  mail_plugins = " quota sieve"
}
protocol imap {
  mail_plugins = " quota imap_quota zlib imap_zlib"
}

# Just querying the setting defaults to the global setting, not protocol scoped:
$ doveconf mail_plugins
mail_plugins =  quota
# The setting for the LMTP protocol:
$ doveconf -f protocol=lmtp
mail_plugins =  quota sieve
```

---

## Reproduction (with working compression)

Take this `compose.yaml` and run the shell commands that follow. No other external files are required, this all works in a self-contained `compose.yaml` example:

```yaml
services:
  dms:
    image: mailserver/docker-mailserver:14.0.0
    hostname: mail.example.test
    # NOTE: Only used for the example for using swaks internally to send mail on port 25
    environment:
      PERMIT_DOCKER: container
    # NOTE: Normally you'd use `volumes` with external files instead of this Compose feature:
    configs:
      - source: dms-accounts
        target: /tmp/docker-mailserver/postfix-accounts.cf
      - source: dovecot-compression
        target: /tmp/docker-mailserver/dovecot.cf

# Using the Docker Compose `configs.content` feature instead of volume mounting separate files.
# NOTE: This feature requires Docker Compose v2.23.1 (Nov 2023) or newer:
# https://github.com/compose-spec/compose-spec/pull/446
# NOTE: `$` literals in `content` are repeated to escape and prevent the `compose.yaml` variable interpolation feature.
configs:
  # DMS expects an account to be configured to run, this example provisions one in advance:
  # Login credentials:
  # user: "john.doe@example.test" password: "secret"
  dms-accounts:
    content: |
      john.doe@example.test|{SHA512-CRYPT}$$6$$sbgFRCmQ.KWS5ryb$$EsWrlYosiadgdUOxCBHY0DQ3qFbeudDhNMqHs6jZt.8gmxUwiLVy738knqkHD4zj4amkb296HFqQ3yDq4UXt8.

  # LMTP protocol for Postfix => Dovecot delivery (enable zlib plugin for on-disk compression)
  # IMAP protocol for MUA (enable imap_zlib plugin for compressed network transfer to MUA client)
  dovecot-compression:
    content: |
      protocol lmtp {
        mail_plugins = $$mail_plugins zlib
      }
      protocol imap {
        mail_plugins = $$mail_plugins imap_zlib
      }

      # On-disk compression:
      plugin {
        zlib_save = gz
        zlib_save_level = 9
      }

      # IMAP network compression:
      plugin {
        imap_compress_deflate_level = 6
      }
```

```console
# Start DMS container and shell into it:
$ docker compose up --force-recreate -d
$ docker compose exec -it dms bash

# Send a mail to the mailbox, it'll be stored compressed:
$ swaks --helo not-relevant.test --server 0.0.0.0 --port 25 --from hello@not-relevant.test --to john.doe@example.test

# Verify compression - GZIP will look like gibberish instead of plain-text output
$ cat /var/mail/example.test/john.doe/new/*

# Verify Dovecot config is correct for `mail_plugins` per protocol:
$ doveconf -f protocol=lmtp mail_plugins
mail_plugins =  quota sieve zlib
$ doveconf -f protocol=imap mail_plugins
mail_plugins =  quota imap_quota imap_zlib
```

**Comment by alex9434** [comment]
@polarathene 
Wow, thank you very much for this very thorough help. I really, really appreciate that! Apologies for the late reply, I was travelling. Based on the evidence you provided, I confirm tha.t this was a mis-configuration on my part. I will try your solution in the next few days and report back.

**Comment by georglauterbach** [comment]
I'm glad you figured this one out @polarathene, nice work!

What's really striking is the amount of user misconfigurations that we see in our issue tracker, really (no offense @alex9434, happens to everyone, and has happened to me too 😄).

**Comment by alex9434** [comment]
@georglauterbach and @polarathene 
I just received the notification for your last comment on github to my email and it is stored compressed (6,699 byte uncompressed and 2572 bytes compressed)! So everything working! I understand that it can be frustrating for you to deal with all these misconfigurations. Its a complex subject with difficult documentation due to the multiple tools involved.

**Comment by georglauterbach** [comment]
> @georglauterbach and @polarathene I just received the notification for your last comment on github to my email and it is stored compressed (6,699 byte uncompressed and 2,572 bytes compressed)! So everything working! I understand that it can be frustrating for you to deal with all these misconfigurations. Its a complex subject with difficult documentation due to the multiple tools involved.

👍🏼 
 
> BTW: If at any point you want to include compression in your documentation, I am happy to write a few paragraphs based on the lessens learned.

That would be very much appreciated!

**Comment by polarathene** [comment]
> However, you seem to actually need zlib in protocol lmtp and protocol imap.

Ah ok yes, that probably makes more sense. The Dovecot docs page was vague with their configuration snippet, but I guess it was intended as a global `mail_plugins` setting so that any other scopes would inherit that.

I've seen a docs page that specifically demonstrates this gotcha with `quota` plugin. However when I used our override support to override the global, that didn't preserve changes DMS configures for the global, which would output a clear warning about that. So I decided to scope it to the `lmtp` protocol. Adding it to `imap` (_and `pop3` if used I assumed_) would also be necessary then 👍 

> BTW: If at any point you want to include compression in your documentation, I am happy to write a few paragraphs based on the lessens learned.

That could be done, or a simpler way is to take the reproduction example I provided earlier and contribute the `.yaml` to our [demo setups](https://github.com/docker-mailserver/docker-mailserver/tree/1a938dfb15e94f43f1af5ddaf7b8276aba92a825/demo-setups) repo folder. If you prefer to make the example multiple files, just group them under a dictionary, name could be something like `dovecot-compression`.

It seems fairly simple, that we can just take that with your fix applied to what I prepared earlier, and you'll see the other examples link to the related issue or discussion for more context. It's not as nice as rendered documentation, but it is somewhat discoverable and very quick to test locally with copy/paste of a single `compose.yaml` and a `docker compose up`. Each current example links to my comments with shell commands to verify and test functionality.

If you would still like to go ahead with a docs contribution, we tend to add these to the ["Examples" section](https://docker-mailserver.github.io/docker-mailserver/v14.0/examples/tutorials/basic-installation/) as community use-cases. If they become more commonly used / relevant they are migrated to the configuration section.
